assembly/src/ast/module.rs

```diff
@@ -1,321 +1,688 @@
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        match self {
+            Self::Library => f.write_str("library"),
+            Self::Executable => f.write_str("executable"),
+            Self::Kernel => f.write_str("kernel"),
         }
-        if reexported_procs.len() > MAX_REEXPORTED_PROCS {
-            return Err(ParsingError::too_many_module_procs(
-                reexported_procs.len(),
-                MAX_REEXPORTED_PROCS,
-            ));
+    }
+}
+
+impl Serializable for ModuleKind {
+    fn write_into<W: ByteWriter>(&self, target: &mut W) {
+        target.write_u8(*self as u8)
+    }
+}
+
+impl Deserializable for ModuleKind {
+    fn read_from<R: ByteReader>(source: &mut R) -> Result<Self, DeserializationError> {
+        match source.read_u8()? {
+            0 => Ok(Self::Library),
+            1 => Ok(Self::Executable),
+            2 => Ok(Self::Kernel),
+            n => Err(DeserializationError::InvalidValue(format!("invalid module kind tag: {n}"))),
         }
-        if let Some(ref docs) = docs {
-            if docs.len() > MAX_DOCS_LEN {
-                return Err(ParsingError::module_docs_too_long(docs.len(), MAX_DOCS_LEN));
-            }
+    }
+}
+
+/// The abstract syntax tree for a single Miden Assembly module.
+///
+/// All module kinds share this AST representation, as they are largely identical. However, the
+/// [ModuleKind] dictates how the parsed module is semantically analyzed and validated.
+#[derive(Clone)]
+pub struct Module {
+    /// The span covering the entire definition of this module
+    span: SourceSpan,
+    /// If available/known, the source contents from which this module was parsed. This is used
+    /// to provide rich diagnostics output during semantic analysis.
+    ///
+    /// In cases where this file is not available, diagnostics will revert to a simple form with
+    /// a helpful message, but without source code snippets.
+    source_file: Option<Arc<SourceFile>>,
+    /// The documentation associated with this module.
+    ///
+    /// Module documentation is provided in Miden Assembly as a documentation comment starting on
+    /// the first line of the module. All other documentation comments are attached to the item the
+    /// precede in the module body.
+    docs: Option<Span<String>>,
+    /// The fully-qualified path representing the name of this module
+    path: LibraryPath,
+    /// The kind of module this represents
+    kind: ModuleKind,
+    /// The imports defined in the module body
+    pub(crate) imports: Vec<Import>,
+    /// The procedures (defined or re-exported) in the module body
+    ///
+    /// NOTE: Despite the name, the procedures in this set are not necessarily exported, the
+    /// individual procedure item must be checked to determine visibility
+    pub(crate) procedures: Vec<Export>,
+}
+
+impl Spanned for Module {
+    fn span(&self) -> SourceSpan {
+        self.span
+    }
+}
+
+/// Construction
+impl Module {
```

Comment:
Yeah, I see no value in the C++ headers either. Not only it is not supported by the rust tooling, but AFAIK it is not usable by the editors either (meaning it is not a collapsible region). And I have the opposite perspective that you have with the headers being misleading, to me it is extra work to maintain them. I'm all in favor for removing it.

Reply:
Also +1 - I never found the headers useful myself (exactly because there's no tool support for them). I also agree that this is a better alternative convention.

Reply:
@bobbinth mentioned to me that he does find the section headers helpful in orienting himself in the code, so we probably should discuss this as a team and see where we land - I certainly don't want to make things harder on anyone. I am curious if there is maybe an alternative way to achieve the same ease of orienteering within the bounds of what our tools support and provide for us, but I think we'll have to discuss to figure that out. 

See [this Discord thread](https://discord.com/channels/893151281848406016/1148602660132487279/1217866991600275557) for discussion on this and other questions of this nature.